solidity/contracts/ValidatorAnnounce.sol

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
+
+// ============ Internal Imports ============
+import {IValidatorAnnounce} from "../interfaces/IValidatorAnnounce.sol";
+import {IMailbox} from "../interfaces/IMailbox.sol";
+import {TypeCasts} from "./libs/TypeCasts.sol";
+// ============ External Imports ============
+import {EnumerableSet} from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+/**
+ * @title ValidatorAnnounce
+ * @notice Stores the location(s) of validator signed checkpoints
+ */
+contract ValidatorAnnounce is IValidatorAnnounce {
+    // ============ Libraries ============
+
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using TypeCasts for address;
+
+    // ============ Constants ============
+
+    // Address of the mailbox being validated
+    address public immutable mailbox;
+    // Domain of chain on which the contract is deployed
+    uint32 public immutable localDomain;
+
+    // ============ Public Storage ============
+
+    // The set of validators that have announced
+    EnumerableSet.AddressSet private validators;
+    // Storage locations of validator signed checkpoints
+    mapping(address => string[]) private storageLocations;
+    // Mapping to prevent the same announcement from being registered
+    // multiple times.
+    mapping(bytes32 => bool) private replayProtection;
+
+    // ============ Events ============
+
+    /**
+     * @notice Emitted when a new validator announcement is made
+     * @param validator The address of the announcing validator
+     * @param storageLocation The storage location being announced
+     */
+    event ValidatorAnnouncement(
+        address indexed validator,
+        string storageLocation
+    );
+
+    // ============ Constructor ============
+
+    constructor(address _mailbox) {
+        mailbox = _mailbox;
+        localDomain = IMailbox(mailbox).localDomain();
+    }
+
+    // ============ External Functions ============
+
+    /**
+     * @notice Announces a validator signature storage location
+     * @param _storageLocation Information encoding the location of signed
+     * checkpoints
+     * @param _signature The signed validator announcement
+     * @return True upon success
+     */
+    function announce(
+        address _validator,
+        string calldata _storageLocation,
+        bytes calldata _signature
+    ) external returns (bool) {
+        // Ensure that the same storage metadata isn't being announced
+        // multiple times for the same validator.
+        bytes32 _replayId = keccak256(
+            abi.encodePacked(_validator, _storageLocation)
```

Comment:
It prevents multiple identical entries in the storageLocations[_validator] array.

Alternatively we could iterate over that array, maybe that would be better..

Reply:
Wdyt?

Reply:
Agreed, also seems unnecessary

Reply:
I feel like it's a good idea to have some replay protection, which approach do you prefer?

Reply:
If you don't have replay protection anyone can just spam this contract and replay announcements

Reply:
talked offline with @asaj
assuming we need to support multiple contemporaneous storage modalities
agree that not requiring validator keys to announce themselves is worth this complexity
otherwise we could use a nonce